apps/performance-tests/src/util/Datasets.ts

```diff
@@ -127,4 +164,203 @@ export class Datasets {
+          }
+        }
+      }
+    });
+
+    console.log(`${numElements} elements: Done.`);
+  }
+
+  private static async createFunctionalElements3dModel(name: string, localPath: string, numElements: number) {
+    console.log(`${numElements} elements: Creating...`);
+    const schema = await this.getSchemaFromPackage("functional-schema", "Functional.ecschema.xml");
+
+    await createIModel(name, localPath, async (builder) => {
+      await builder.importFullSchema(schema);
+      const { id: physicalModelId } = insertPhysicalModelWithPartition({ builder, codeValue: "test physical model" });
+      const { id: functionalModelId } = insertFunctionalModelWithPartition({ builder, codeValue: "test functional model" });
+      const { id: categoryId } = insertSpatialCategory({ builder, codeValue: "test category" });
+
+      const numberOfGroups = 1000;
+      const elementsPerGroup = numElements / numberOfGroups;
+
+      for (let i = 0; i < numberOfGroups; ++i) {
+        let physicalElementParentId: string | undefined;
+        let functionalElementParentId: string | undefined;
+
+        for (let j = 0; j < elementsPerGroup; ++j) {
+          physicalElementParentId = insertPhysicalElement({
+            builder,
+            parentId: physicalElementParentId,
+            modelId: physicalModelId,
+            categoryId,
+          }).id;
+          functionalElementParentId = insertFunctionalElement({
+            builder,
+            parentId: functionalElementParentId,
+            modelId: functionalModelId,
+            representedElementId: physicalElementParentId,
+            relationshipName: "PhysicalElementFulfillsFunction",
+          }).id;
+        }
+      }
+    });
+
+    console.log(`${numElements} elements: Done.`);
+  }
+
+  private static async createNestedFunctionalElements2dModel(name: string, localPath: string, numElements: number) {
+    console.log(`${numElements} elements: Creating...`);
+    const schema = await this.getSchemaFromPackage("functional-schema", "Functional.ecschema.xml");
+
+    await createIModel(name, localPath, async (builder) => {
+      await builder.importFullSchema(schema);
+
+      const { id: drawingModelId } = insertDrawingModelWithPartition({ builder, codeValue: "test drawing model" });
+      const { id: functionalModelId } = insertFunctionalModelWithPartition({ builder, codeValue: "test functional model" });
+      const { id: categoryId } = insertDrawingCategory({ builder, codeValue: "test drawing category" });
+      const { id: graphicsElementId } = insertDrawingGraphic({ builder, modelId: drawingModelId, categoryId });
+
+      const numberOfGroups = 1000;
+      const elementsPerGroup = numElements / numberOfGroups;
+
+      for (let i = 0; i < numberOfGroups; ++i) {
+        let functionalElementParentId: string | undefined;
+        let graphicsElementParentId = insertDrawingGraphic({
+          builder,
+          categoryId,
+          modelId: drawingModelId,
+        }).id;
```

Comment:
why is this top-level element created outside the loop?

Reply:
Refactored the entire functional 2D iModel setup, to make it more clear.